### ISSUE TITLE
Shouldn't the dice coefficient loss be negative?

### ISSUE THREAD

**Issue by woodywff** [IAmSuyogJadhav/3d-mri-brain-tumor-segmentation-using-autoencoder-regularization#7]
The loss function in this code is:
```
def loss_(y_true, y_pred):
        y_true_f = K.flatten(y_true)
        y_pred_f = K.flatten(y_pred)
        intersection = K.sum(K.abs(y_true_f * y_pred_f), axis=-1)
        loss_dice = (2. * intersection) / (
            K.sum(K.square(y_true_f), -1) + K.sum(K.square(y_pred_f), -1) + e)

        return loss_dice + 0.1 * loss_L2 + 0.1 * loss_KL
```
But I think our target is to maximize the loss_dice part, isn't it?<p> Are we missing a minus here? It goes like:
`return - loss_dice + 0.1 * loss_L2 + 0.1 * loss_KL`

Thanks :-)

**Comment by IAmSuyogJadhav** [comment]
I am not much familiar with the concept of the dice loss as a whole. Intuitively, I think there should be a minus there, but on searching the web for a lot of time, this was the only implementation I found over and over again. I am still a bit skeptical about it.
Can you please confirm if adding the minus sign is not making things worse? I will then add it to the main code. If you have used the dice loss earlier and know for sure that we need to add the minus, let me know as well.
Awaiting your reply.

**Comment by doc78** [comment]
I agree with woodywff, a minus is missing there since the score of your dice loss increases when the differences between prediction and label are bigger, and we want to reduce that difference. So the correct formula is:

return - loss_dice + 0.1 * loss_L2 + 0.1 * loss_KL

I trained the network and I got good results adding the minus term. I also tested without the minus and I got obviously wrong predictions.
BTW I found that even into original paper the minus is missing.

There is something more I would like to investigate: the loss_KL term is too small compared to others loss term. However, I tryed to avoid dividing by the number of voxels and the model was too much instable. So I'm trying to get a good compromise between the terms. As I get the results I'll let you know.

**Comment by IAmSuyogJadhav** [comment]
Alright, did that here (a93de7c95d95f9ce4d0d65d11183edeb6eeab133). And yes @doc78 , the original author did leave out the minus sign in his paper. He could have called it _coefficient of dice loss_ (which is basically negative of the dice loss) to reduce confusion. That formula in the paper was one of the reasons I did not have a minus in the implementation here.
Btw, regarding the weights of different loss terms: I took the ones used by the author. I think value of those terms should vary according to the task. I will add another argument in the `build_model` function to allow for different weights, so that you can experiment easily.

**Comment by IAmSuyogJadhav** [comment]
I have implemented the customizable loss weights in this one (85db9debe14afb5588d7ef22ebb5b51f8ab13982 ). Currently, I am away from my PC, will merge that later once I test there are no issues. You can use that branch if you want, for now.

**Comment by IAmSuyogJadhav** [comment]
Closing this issue since the OP's problem has been resolved. Feel free to open again, if needed.
